### PR TITLE
Templates for memory, communication and Dirac operator

### PR DESCRIPTION
**Pull request by sofiemartins**
* Improved testing (specifically for GPU)
* Reorganized memory, communication and Dirac operator code as templates
* Renaming communication functions everywhere (this is why so many files changed)
* Fixed stride gpu geometry

### CODE REVIEW

File: LibHR/Geometry/communications.c [hunk 1]
```diff
@@ -1,1103 +1,150 @@
...
-
+#include "io.h"
+#include "random.h"
+
+// TODO: put gpu as last suffix
+// TODO: fill buffers needs gpu suffix
+// TODO: MPI error management
+    
+#define random_double ranlxd
+#define random_float ranlxs
```
Comment (claudiopica) [inline LibHR/Geometry/communications.c:16]: what are these for?

File: LibHR/Geometry/TMPL/communications.c.tmpl [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+* Copyright (c) 2023, Sofie Martins                                         *   
+* All rights reserved.                                                      * 
+\***************************************************************************/
+
+#include <string.h>
+#include "geometry.h"
+#include "Utils/generics.h"
+
+#if !defined _FIELD_NAME
+#error Missing _FIELD_NAME in communications_gpu.c
```
Comment (claudiopica) [inline LibHR/Geometry/TMPL/communications.c.tmpl:11]: change file name

File: LibHR/IO/archive_spinors.c [hunk 3]
```diff
@@ -10,15 +10,15 @@
...
+#include "libhr.h"
+/*#include "global.h"
 #include "Geometry/communications.h"
 #include "Utils/boundary_conditions.h"
 #include "Utils/timing.h"
 #include "Observables/avr_plaquette.h"
 #include "Random/ranlux.h"
 #include "Inverters/linear_algebra.h"
-#include "io.h"
+#include "io.h"*/
```
Comment (claudiopica) [inline LibHR/IO/archive_spinors.c:21]: just remove the commented headers

File: LibHR/Memory/TMPL/field_alloc.c.tmpl [hunk 4]
```diff
@@ -0,0 +1,189 @@
+/***************************************************************************\
+* Copyright (c) 2008, 2022, Claudio Pica, Sofie Martins                     *   
+* All rights reserved.                                                      * 
+\***************************************************************************/
+
+// TODO: Deallocate sendbuffers
```
Comment (claudiopica) [inline LibHR/Memory/TMPL/field_alloc.c.tmpl:6]: is this an old comment? I think we already do that now

File: LibHR/Random/random_spinor_field.c [hunk 5]
```diff
@@ -16,7 +17,12 @@ void gaussian_spinor_field(spinor_field *s)
 		gauss((double *)(s->ptr + (type->master_start[i] - type->master_shift)), (type->master_end[i] - type->master_start[i] + 1) * sizeof(suNf_spinor) / sizeof(double));
 
 	spinor_field_mul_f_cpu(s, c1, s);
-	apply_BCs_on_spinor_field(s);
```
Comment (claudiopica) [inline LibHR/Random/random_spinor_field.c:19]: why is this removed?

File: Make/MkFlags.ini [hunk 6]
```diff
@@ -14,6 +14,7 @@ MACRO += UPDATE_EO
 MACRO += NDEBUG
 MACRO += CHECK_SPINOR_MATCHING
 MACRO += IO_FLUSH 
+MACRO += DPHI_FLT
```
Comment (claudiopica) [inline Make/MkFlags.ini:17]: what is this? It's better to not commit changes to this file

File: Make/MkFlags.ini [hunk 7]
```diff
@@ -1,11 +1,11 @@
 NG = 2
 REPR = REPR_FUNDAMENTAL
-#REPR = REPR_ANTISYMMETRIC
 GAUGE_GROUP = GAUGE_SUN
 #MACRO += WITH_MPI 
 #MACRO += WITH_GPU
 
-MACRO += WITH_NEW_GEOMETRY
+#MACRO += WITH_NEW_GEOMETRY
+#MACRO += FIXED_STRIDE
```
Comment (claudiopica) [inline Make/MkFlags.ini:8]: Did you add this option to write_mkflags?

File: TestProgram/Benchmarks/speed_test_diracoperator_gpu.c [hunk 8]
```diff
@@ -36,10 +38,10 @@ int main(int argc,char *argv[])
...
   gaussian_spinor_field(s1);
   lprintf("LA TEST",0,"un sito %lf\n",creal(s0->ptr[2].c[0].c[0]));
-  copy_to_gpu_spinor_field_f(s0);
-  copy_to_gpu_spinor_field_f(s1);
-  copy_to_gpu_spinor_field_f(s2);
-  copy_to_gpu_spinor_field_f(s3);
+  //copy_to_gpu_spinor_field_f(s0);
+  //copy_to_gpu_spinor_field_f(s1);
+  //copy_to_gpu_spinor_field_f(s2);
+  //copy_to_gpu_spinor_field_f(s3);
```
Comment (claudiopica) [inline TestProgram/Benchmarks/speed_test_diracoperator_gpu.c:44]: remove if not needed

File: TestProgram/Benchmarks/speed_test_reading_gpu.c [hunk 9]
```diff
@@ -0,0 +1,39 @@
+/*******************************************************************************
+*
+* NOCOMPILE= !WITH_GPU
```
Comment (claudiopica) [inline TestProgram/Benchmarks/speed_test_reading_gpu.c:3]: shouldn't this test work also on CPU?

File: TestProgram/Geometry/check_gpu_geometry.in [hunk 10]
```diff
@@ -1,8 +1,8 @@
 // Global variables 
 GLB_T = 8 
-GLB_X = 4
-GLB_Y = 4 
-GLB_Z = 4
+GLB_X = 7
+GLB_Y = 7
+GLB_Z = 7
```
Comment (claudiopica) [inline TestProgram/Geometry/check_gpu_geometry.in:5]: this should be even

File: TestProgram/LinearAlgebra/check_gpu_2.in [hunk 11]
```diff
@@ -1,8 +1,8 @@
 // Global variables 
 GLB_T = 16 //Global T size
-GLB_X = 16
-GLB_Y = 16
-GLB_Z = 16
+GLB_X = 17
+GLB_Y = 17
+GLB_Z = 17
```
Comment (claudiopica) [inline TestProgram/LinearAlgebra/check_gpu_2.in:5]: should be even

**Review by claudiopica — CHANGES_REQUESTED** [review]
Hi Sofie, very good. there are a few minor things.
Regarding the header files:
1) do not create headers that do not correspond to c files;
2) make sure you reference the implementation file in the header file so that the reader knows where each declared function is defined.

File: Include/Geometry/strided_reads_gpu.h [hunk 12]
```diff
@@ -0,0 +1,180 @@
...
+#ifdef FIXED_STRIDE
+    #define THREADSIZE 32
+#else
+    #define THREADSIZE 1
+#endif
+
+//#include "libhr_core.h"
+#include "geometry.h"
+
+#ifdef __cplusplus
```
[the remark this answers is not in the document]
Comment (sofiemartins) [reply Include/Geometry/strided_reads_gpu.h:13]: These are the only gpu read/write functions that are used at the moment. They are directly used kernels, such as the dirac operator kernels, syncing the field to the sendbuffer in the new geometry and linear algebra. The reading and writing functions in the template that can also be used in C code actually call these C++ functions. 

I am planning to remove the old generated versions in Include/Geometry/strided_reads.h but was reluctant to do it now, because of the amount of changes that I already made.

File: LibHR/Geometry/communications.c (hunk 1 above)
Comment (sofiemartins) [reply LibHR/Geometry/communications.c:16]: Just so that in the template below I can use the real type given as _REAL to construct the right function name. These are not central to the communications but for the fill buffer functions, which, at the moment, I only use for testing. 

This is line 171 in the corresponding template. They are, however, supposed to be only local, so it would probably be best if I moved this into the template and undefine at the end of the file.

File: Make/MkFlags.ini (hunk 6 above)
Comment (sofiemartins) [reply Make/MkFlags.ini:17]: This is a necessary change. The single precision gauge field u_gauge_f_flt is only allocated in the process_init.c if this compile flag is defined. I had some trouble with this, because for example check_diracoperator_6.c depends on this gauge field being allocated but so far there was no NOCOMPILE to prevent the check from being run without it (I added it now). It is clear, that we don't want to allocate another gauge field if we are not sure that we want to use it. But it is a little intransparent, that you cannot run anything that needs the single precision dirac operator without having added this compile flag. I would keep it there, but maybe disable it. Or by default enable it, so that people don't get cryptic segfaults.

File: TestProgram/Benchmarks/speed_test_reading_gpu.c (hunk 9 above)
Comment (sofiemartins) [reply TestProgram/Benchmarks/speed_test_reading_gpu.c:3]: It does but it is an unfinished draft for a GPU memory access speed test, that I intent to change to something better. If its a problem that this is there, I will remove this file for now. If we want a linear algebra cpu/gpu comparison speed test, this should be maybe designed differently.

File: Include/Update/dirac.h [hunk 13]
```diff
@@ -15,49 +15,19 @@
...
+#include "Utils/generics.h"
 
 #ifdef __cplusplus
     extern "C" {
 #endif
 
-#ifdef __cplusplus
-//there is no 'restrict' in C++, use an extension
-#define restrict __restrict__
-#endif
```
[the remark this answers is not in the document]
Comment (sofiemartins) [reply Include/Update/dirac.h:26]: It is not removed, I just moved in into the template dirac.h.tmpl. Is this needed globally?

File: Include/Geometry/strided_reads_gpu.h (hunk 12 above)
[the remark this answers is not in the document]
Comment (claudiopica) [reply Include/Geometry/strided_reads_gpu.h:13]: It'd be better to move the templated c++ implementation of the functions to a separate hpp file in LibHR, and only keep the c functions in this header

File: Include/Update/dirac.h (hunk 13 above)
[the remark this answers is not in the document]
Comment (claudiopica) [reply Include/Update/dirac.h:26]: yes it is, especially here because some functions are not in the template.

File: LibHR/Geometry/communications.c (hunk 1 above)
Comment (claudiopica) [reply LibHR/Geometry/communications.c:16]: ok I see. I think it is fine for now, but we should really use _Generic for this.

File: Make/MkFlags.ini (hunk 6 above)
Comment (claudiopica) [reply Make/MkFlags.ini:17]: ok, you are right, this is not good. we should improve it. This macro shouldn't be necessary at all.
Let's keep it as it is for this PR, but please don't commit the ini file.

File: TestProgram/Benchmarks/speed_test_reading_gpu.c (hunk 9 above)
Comment (claudiopica) [reply TestProgram/Benchmarks/speed_test_reading_gpu.c:3]: ok. no problem.

File: LibHR/Memory/TMPL/field_alloc.c.tmpl [hunk 14]
```diff
@@ -0,0 +1,189 @@
...
+    *f = amalloc(n * sizeof(_FIELD_TYPE), ALIGN);
+    error(f == NULL, 1, __func__, "Could not allocate memory space for field structure");
+    for (int i = 0; i < n; ++i) { (*f)[i].type = type; }
+}
+
+_DECLARE_STATIC(allocate_cpu_field_data_,(unsigned int n, _FIELD_TYPE *f, geometry_descriptor *type)) {
+    if (alloc_mem_t & CPU_MEM) {
+        int bytes_per_site = sizeof(_SITE_TYPE);
+        int number_of_sites = n * _FIELD_DIM * _NUMBER_OF_SITES(_GEOM_TYPE);
+        int field_size  = bytes_per_site * number_of_sites;
```
Comment (claudiopica) [inline LibHR/Memory/TMPL/field_alloc.c.tmpl:61]: these int variables overflow for large lattices. can you declare them as size_t ?
Same below for GPU code

**Review by claudiopica — APPROVED** [review]
All looks good. great work!
